lib/src/conflicts.rs

```diff
@@ -483,30 +499,51 @@ pub async fn update_from_content(
 
     // If the user edited the empty placeholder for an absent side, we consider the
     // conflict resolved.
-    if zip(contents.iter(), file_ids.iter())
+    if zip(contents.iter(), used_file_ids.iter())
         .any(|(content, file_id)| file_id.is_none() && !content.is_empty())
     {
         let file_id = store.write_file(path, &mut &content[..])?;
         return Ok(Merge::normal(file_id));
     }
 
     // Now write the new files contents we found by parsing the file with conflict
-    // markers. Update the Merge object with the new FileIds.
-    let builder: BackendResult<MergeBuilder<Option<FileId>>> =
-        zip(contents.iter(), file_ids.iter())
-            .map(|(content, file_id)| {
-                match file_id {
-                    Some(_) => {
-                        let file_id = store.write_file(path, &mut content.as_slice())?;
-                        Ok(Some(file_id))
-                    }
-                    None => {
-                        // The missing side of a conflict is still represented by
-                        // the empty string we materialized it as
-                        Ok(None)
-                    }
+    // markers.
+    let mut new_file_ids: Vec<Option<FileId>> = zip(contents.iter(), used_file_ids.iter())
+        .map(|(content, file_id)| -> BackendResult<Option<FileId>> {
+            match file_id {
+                Some(_) => {
+                    let file_id = store.write_file(path, &mut content.as_slice())?;
+                    Ok(Some(file_id))
                 }
-            })
+                None => {
+                    // The missing side of a conflict is still represented by
+                    // the empty string we materialized it as
+                    Ok(None)
+                }
+            }
+        })
+        .try_collect()?;
+
+    // If the conflict was simplified, expand the conflict to the original
+    // number of sides.
+    if new_file_ids.len() != file_ids.iter().len() {
+        assert_eq!(new_file_ids.len(), simplified_file_ids.iter().len());
+        let simplified_file_ids_index_map: HashMap<_, _> = simplified_file_ids
+            .iter()
+            .enumerate()
+            .map(|(index, file_id)| (file_id, index))
             .collect();
```

Comment:
Will this be correct if the simplified conflict had repeated terms and the user modified only one of them?

Reply:
Yeah, I think this behavior isn't fully accurate. Say the following conflict `A - A + A - B + A` was simplified into `A - B + A`, and then modified into `A1 - B + A2`. What would the expected unsimplified conflict be? Should we just match up `A1` and `A2` to the first two matching `A` terms in the simplified version, then keep any original remaining `A` terms?

Eg. `A1 - B + A2` is expanded back to `A1 - A + A2 - B + A`.

Reply:
Alternatively, we could have `simplify` (or a function based off it) return a list of removed indices which we can use to preserve the original terms at those indices, instead of manually matching them back up.

Reply:
Good point, Martin!

> Eg. A1 - B + A2 is expanded back to `A1 - A + A2 - B + A`.

I think this would be fine. There wouldn't be much harm if it expanded to `A1 - A1 + A2 - B + A1` either, but leaving A-s seems more symmetric.

------

This made me wonder about potential test cases to add. I think `A - A + A - B + A` might actually simplify all the way to `A` due to `A - B + A = A` rule, in which case the issue Martin pointed out will only appear for `A - A + A - B + A - B + C` (or, maybe `D - D + A - B + A - B + C` is a bit clearer). That's a bit of a mouthful to write a test for, but we might need to.

> Should we just match up A1 and A2 to the first two matching A terms in the simplified version, then keep any original remaining A terms?

I'm worried this won't work in the `A - A + A - B + A - B + C` example, will it?

> Alternatively, we could have simplify (or a function based off it) return a list of removed indices which we can use to preserve the original terms at those indices, instead of manually matching them back up.

I don't have a better idea at the moment.

Reply:
Instead of returning indices to delete, `simplify_and_return_mapping` could return the simplified conflict and the mapping of simplified indices to original indices. You'd need to compute these in any case.

Reply:
> > Should we just match up A1 and A2 to the first two matching A terms in the simplified version, then keep any original remaining A terms?
> 
> I'm worried this won't work in the `A - A + A - B + A - B + C` example, will it? If I'm following, it will try to replace the first two `A`-s which are the ones that shouldn't actually be touched.

Matching also preserves whether the term is a add/remove, so it would be `A1 - A + A2 - B + A - B + C` in this example.

Reply:
I guess you are right, we do allow reordering of positive terms. It might make the conflict look slightly different, but we do consider it equivalent and it should hopefully simplify the same. So, as long as we change one positive term A -> A1 and another A-> A2 (and leave the rest), it should be OK.

Reply:
In that case, I'm wondering whether this idea and your code can be made more readable by considering adds and removes separately or adding add/remove to the file ids in the mappings. I'd like to avoid modular arithmetic if possible, and make it more obvious whether or not `.position(...).unwrap()` will ever panic. Are you still polishing it, or should I think about how to clarify it?

Reply:
(Thanks for updating it so fast, by the way!)

Reply:
> I think `A - A + A - B + A` might actually simplify all the way to `A` due to `A - B + A = A` rule

We don't apply that when simplifying (`Merge::simplify()`), only when trying to fully resolve (`Merge::resolve_trivial()`).

Reply:
> What would the expected unsimplified conflict be? Should we just match up `A1` and `A2` to the first two matching `A` terms in the simplified version, then keep any original remaining `A` terms?

SGTM.

> Alternatively, we could have `simplify` (or a function based off it) return a list of removed indices which we can use to preserve the original terms at those indices, instead of manually matching them back up.

Yet another option is to hide this logic inside the `Merge` type by have something like `unsimplified.update_from_simplified(&simplified)` or `simplified.unsimplify(&unsimplified)`. That should be easy to test with unit tests in `merge.rs`. Just a thought; not saying you have to do it that way.

Reply:
I've updated `Merge` to have a separate function to generate the mapping from unsimplified to simplified, and added the `update_from_simplified` method to perform this as well.

Reply:
> I've updated `Merge` to have a separate function to generate the mapping from unsimplified to simplified, and added the `update_from_simplified` method to perform this as well.

Nice, thank you!

> > I think `A - A + A - B + A` might actually simplify all the way to `A` due to `A - B + A = A` rule
> 
> We don't apply that when simplifying (`Merge::simplify()`), only when trying to fully resolve (`Merge::resolve_trivial()`).

(No immediate action needed) This should actually be fixed, I think (in a separate PR; we can leave a TODO or I can file an issue later). Let's say we have

```
B1  B1    D_conflict2
 \/
  A1
``

where `D_conflict2` has a conflict in `file2`, while a merge of the two `B1`s would have a "B1 + B1 - A1" situation in `file1`.

Now, if we merge the two B1s and `D_conflict2`, you'll end up with a conflict in `file1`. If you merged jsut the two B1s, there would be no conflict in file1, so merging the result with `D_conflict2` wouldn't create a conflict there either.

Some degree of non-asociativity of merging is unavoidable with the `B + B - A = B` rule, but I think in this example, it would be quite unexpected that a conflict in file2 affects conflict resolution in file1.

Reply:
I think we already resolve such conflicts here: https://github.com/martinvonz/jj/blob/4cca84515a1c18a221cb2ee31a036bde60a3db2f/lib/src/merged_tree.rs#L466-L475

Reply:
(No action needed) I see, so a `A - A + A - B + A` conflict should never appear in `update_from_content`. It again seems that the easiest way to test the issue Martin pointed out would be either `A - A + A - B + A - B + C` or `D - D + A - B + A - B + C`. Brr. I previously thought that we should test this, but I hereby retract that statement. Functional tests on the `update_from_simplified` will hopefully be good enough.